src/event/reader.rs

```diff
@@ -107,6 +107,15 @@ pub struct EventReader {
     rg_state: Arc<Mutex<ReaderGroupState>>,
 }
 
+impl Drop for EventReader {
+    /// Destructor for reader invoked. This will automatically invoke reader_offline().
+    fn drop(&mut self) {
+        info!("reader {} is dropped", self.id);
+        let runtime = self.factory.runtime().handle().clone();
+        runtime.block_on(self.reader_offline());
```

Comment:
Yeah I don't think we can call `block_on` in `Drop`, this will prevent `EventReader` from being used in async contexts.
I once tried to use channel to `blocking_send` a message in `drop` and it panicked.

Reply:
This is tricky as I couldn't come up with a way to use async function in `Drop`. Perhaps eventually we still need to ask user to call offline manually?

Reply:
You cannot `,await` inside of a drop.

Reply:
I think this problem is causing the build failure.

Reply:
@Tristan1900 found this: https://docs.rs/tokio/1.6.0/tokio/task/index.html#block_in_place
which sounds like the solution we are looking for.

Reply:
probably we want to use this 
https://docs.rs/tokio/1.6.0/tokio/task/index.html#block_in_place

Reply:
Based on what we talked about in the meeting:
The part of `reader_offline` that needs to be async is:
```
        self.rg_state
            .lock()
            .await
            .remove_reader(&self.id, offset_map)
            .await
            .expect("Update ReaderGroup to ensure reader is offline");
```
I think we can factor this out into a function which `reader_offline` can also call. Then we can do the sync work here (maybe also a private function up to you) and pass the map to the async method inside a `spawn`. The map can be moved, and the `rg_state` is an arc, so it can just be cloned.

Reply:
I have updated the code with the solution.